### ISSUE TITLE
ngSubmit should trigger after Asynchronous Validation completes

### ISSUE THREAD

**Issue by DibyodyutiMondal** [angular/angular#31021]
# 🚀 feature request

### Relevant Package

 This feature request is for @angular/forms
Specifically, Reactive Forms

### Description

```
**in Template**
<form [formGroup]="myFormGroup" (ngSubmit)="mFG_submit()"></form>

**In Component**
mFG_submit() {
   if (!myFormGroup.valid)
      return;

   //logic that will run if and only if the entire form is valid after both 
      sync and async validators have finished executing
}
```
The problem with the above code is that if myFormGroup has any controls with asynchronous validators that, the ngSubmit is triggered after the async validators have started executing, but before they have finished executing. As as result, when the 'if' gate inside mFG_submit() is hit, it returns immediately, because the form status is 'PENDING'. As a result, even if the asynchronous validator returns true, i,e, the control is valid, the submission logic is never executed.

### Describe the solution you'd like
 
One solution is to check on every submit if the form status is "PENDING", and then listen to status changes, and emit ngSubmit again, the moment status changes. I would like to have such behaviour provided by the framework
Currently, I use this directive 

```
@Directive({
    selector: 'form[formGroup]'
})
export class ResubmitIfPendingDirective {
    constructor(
        private fgd: FormGroupDirective
    ) {
        this.subscriptions.add(this.resubmission_sub);
    }
    private subscriptions: Subscription = new Subscription();
    private resubmission_sub: Subscription = new Subscription();

    ngOnInit() {
        //listen to ngSubmit of the form
        this.subscriptions.add(this.fgd.ngSubmit.subscribe(() => {

            //if you are already subscribed to status changes, unsubscribe
            this.resubmission_sub.unsubscribe();

            //if your form is pending when submitted, subscribe to status changes
            if (this.fgd.control.pending) {
                this.resubmission_sub = this.fgd.control.statusChanges.pipe(

                   //don't do anything if status changes to PENDING 
                    //(in case of multiple async validators)
                    //Actually I'm not sure if status changes after every single 
                    //async validator or after all async validators are done
                    //so this is just to stay safe
                    //Would be great if the Angular Team could clear my doubt on this
                    takeWhile(() => !this.fgd.control.pending),

                    //status is now changed to non-pending, resubmit
                    first()
                ).subscribe(() => this.fgd.ngSubmit.emit());
            }
        }));
    }

    ngOnDestroy() {
        //stop listening to ngSubmit and status changes
        this.subscriptions.unsubscribe();
    }
}
```

**Comment by jasonaden** [comment]
Interesting issue. This seems to make sense.

**Comment by velraman** [comment]
@jasonaden  Can I Start working in this ?. Thanks!

**Comment by DibyodyutiMondal** [comment]
As you can see, I have added the 'resubmission_sub' to the parent 'subscriptions' in the constructor. However, throughout the lifetime of the directive, the 'resubmission_sub' variable is being assigned a brand new subscription object. After that reassignment, is it enough to unsubscribe to the parent subscription in 'ngOnDestroy'? Or do I have to call unsubscribe on resubmission_sub as well? Can someone please clear this up for me? I am not very knowledgeable on subscriptions.

Another doubt: Is statusChanges emitted after every single async validator completes? or just once after all of them have finished?

**Comment by DibyodyutiMondal** [comment]
So I have been doing a bit of thinking and I realized that there are probably several use cases where people may want to do stuff onSubmit which does not depend on the form validation.

So I propose that the FormGroupDirective have 2 EventEmitters:
- ngSubmit (the current implementation without changes)
- ngSubmitPostValidation

Basically, the idea is that ngSubmit is emitted every time the form is submitted. Once ngSubmit is emitted, you check if the form status is 'PENDING'. If not, you emit ngSubmitPostValidation. If form status is PENDING, then we listen to status changes, and the moment it is no longer pending, we emit ngSubmitPostValidation.

This allows users a freedom to hook into the form submission immediately, or wait till async validation completes, or even both if necessary

**Comment by george43g** [comment]
I have also encountered this issue. It seems that if you have a long form that it's not so much of an issue - however if you're using a reactive form for a simple login, that's when it becomes a problem. Because a user presses enter to login after they type their password. This triggers the validation, but because the state is "PENDING", the submission cannot go ahead. The solution is to set up a complex observable that waits until the status is valid etc...

A cleaner solution would be much appreciated.

**Comment by DibyodyutiMondal** [comment]
Yes, that is exactly the situation that made me pursue this route. XD

**Comment by cah-chase-spencer** [comment]
Ran into this today. This would be awesome!

**Comment by DibyodyutiMondal** [comment]
My original recommended solution had a mistake in the code, I have now updated it

I had used `takewhile()` as the first pipe operator, which is completely wrong.
To get desired behaviour it should actually be `filter()`

**Comment by DibyodyutiMondal** [comment]
Is there any update on this?

**Comment by arjunasuresh3** [comment]
1+

**Comment by Sirstep** [comment]
+1

**Comment by IT-CASADO** [comment]
+1

**Comment by manju-reddys** [comment]
+1

**Comment by dylhunn** [comment]
It's pretty odd that this works the way it does. We should add a flag on `ngForm` to enable waiting for async validators, or perhaps add a new emitter.

**Comment by DibyodyutiMondal** [comment]
I don't think a new emitter would be necessary, and following is why I think not.

In my experience, any ui-related behaviour, whether simple or complex can be made by piping the form group's status-change or value-change  observables, or to use the observables from the form controls, or even using the (click) handler of the submit button or a combination of the above. It does not need the ngSubmit observable at all. 

Thus, I believe we can leave ui changes out of this discussion. That leaves only the data processing. I know this is rather opinionated of me, but I do find this separation of concerns useful.

Carrying on, since we use ngSubmit to start processing the values in the form controls, ideally, one would expect validators to have run by the time it is emitted and a clear yes/no answer  to the question, "are the values valid? Should I process the data or ignore it?"

Therefore, I dare say that for most use-cases having a flag 'waitForAsyncValidation' which defaults to `true` would be the way to go. 

Additionally, while it is not required, we could also add another flag: 'submitWhenValid'. Whether it should default to `true` or `false`, I leave up to you to decide. Like I said, it's not necessary, as this can easily be done with a simple 'if not valid then return' 2-liner. But it would be nice to know what you think.

**Comment by dylhunn** [comment]
> for most use-cases having a flag 'waitForAsyncValidation' which defaults to true would be the way to go.

I agree that this is the easiest solution. However, we'd need to default to `false` to avoid making breaking changes. Then we could add a migration and flip the default in a future framework version.

Would you be comfortable putting together a PR to fix this? I'd like to fix it, but don't have the bandwidth right now. Happy to review and merge it though.

**Comment by DibyodyutiMondal** [comment]
Ah yes, I didn't think about that. My bad.

**Comment by sasunstepanyan777** [comment]
I think we should add that flag as option into formBuilder.group like:
`form: FormGroup = this.formBuilder.group({
    ...control
  }, {
    waitForAsyncValidation: true
  });`

and have a method in FormGroup to change this flag

**Comment by buddhiko1** [comment]
Any updates on this?

**Comment by kashcode** [comment]
Any updates on this?
